File: ABC/ABC005/D_100.cpp
// 満点解法 累積和 前計算O(N^2) クエリあたりO(N^4) → O(N^2+N^4・Q)

#include <bits/stdc++.h>
using namespace std;

int main() {
    int n, q, p;
    cin >> n;
    vector<vector<int>> d(n, vector<int>(n));
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            cin >> d[i][j];
        }
    }
    int dac[51][51];
    for (int i = 0; i < n; i++) {
        for (int j = 0; j < n; j++) {
            dac[i + 1][j + 1] = dac[i + 1][j] + d[i][j];
        }
    }
    for (int i = 0; i < n; i++) {
        for (int j = 0; j <= n; j++) {
            dac[i + 1][j] += dac[i][j];
        }
    }
    cin >> q;
    while (cin >> p) {
        int ans = 0;
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= n; j++) {
                if (i * j > p)
                    break;
                int now = 0;
                for (int s = 0; s <= n - i; s++) {
                    for (int t = 0; t <= n - j; t++) {
                        now = max(now, dac[s][t] + dac[s + i][t + j] - dac[s][t + j] - dac[s + i][t]);
                    }
                }
                ans = max(ans, now);
            }
        }
        cout << ans << '\n';
    }
}

File: Python/ABC/ABC005/D_100.py
n = int(input())
d = [list(map(int, input().split())) for _ in range(n)]
dac = [[0] * (n + 1) for _ in range(n + 1)]
for i in range(n):
    for j in range(n):
        dac[i + 1][j + 1] = dac[i + 1][j] + d[i][j]
for i in range(n):
    for j in range(n):
        dac[i + 1][j] += dac[i][j]
for _ in range(int(input())):
    p = int(input())
    ans = 0
    for i in range(1, n + 1):
        for j in range(1, n + 1):
            if i * j > p:
                break
            now = 0
            for s in range(n - i + 1):
                for t in range(n - j + 1):
                    now = max(now, dac[s][t] + dac[s + i][t + j] - dac[s][t + j] - dac[s + i][t])
            ans = max(ans, now)
    print(ans)